File: code/javascript/exercises.js
import fs from "fs";

export function firstThenApply(strings, predicate, fn) {
  const hit = Array.isArray(strings)
    ? strings.find((s) => (predicate ? predicate(s) : false))
    : undefined;

  return hit !== undefined ? (fn ? fn(hit) : undefined) : undefined;
}

export function* powersGenerator({ ofBase: base, upTo: limit }) {
  let v = 1;
  while (v <= limit) {
    yield v;
    v *= base;
  }
}

export function meaningfulLineCount(filename) {
  const text = fs.readFileSync(filename, "utf8");
  return text
    .split(/\r?\n/)
    .filter((line) => {
      const s = line.replace(/^\s+/, "");
      if (s === "" || /^\s*$/.test(line)) return false;
      if (s.startsWith("#")) return false;
      return true;
    }).length;
}

export function say(word) {
  const words = [];
  if (word === undefined) return "";
  words.push(word);

  function inner(...nextArgs) {
    if (nextArgs.length === 0) {
      return words.join(" ");
    }
    const [next] = nextArgs;
    words.push(next);
    return inner;
  }

  return inner;
}

export class Quaternion {
  constructor(a = 0, b = 0, c = 0, d = 0) {
    this.a = +a;
    this.b = +b;
    this.c = +c;
    this.d = +d;
    Object.freeze(this);
  }

  plus(o) {
    return new Quaternion(
      this.a + o.a,
      this.b + o.b,
      this.c + o.c,
      this.d + o.d
    );
  }

  times(o) {
    const { a: a1, b: b1, c: c1, d: d1 } = this;
    const { a: a2, b: b2, c: c2, d: d2 } = o;
    return new Quaternion(
      a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
      a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
      a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
      a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2
    );
  }

  get coefficients() {
    return [this.a, this.b, this.c, this.d];
  }

  get conjugate() {
    return new Quaternion(this.a, -this.b, -this.c, -this.d);
  }

  equals(o) {
    return (
      this.a === o.a &&
      this.b === o.b &&
      this.c === o.c &&
      this.d === o.d
    );
  }

  toString() {
    const parts = [];
    if (this.a !== 0) parts.push(`${this.a}`);

    const add = (c, s) => {
      if (c === 0) return;
      const sign = c > 0 ? "+" : "-";
      const mag = Math.abs(c);
      const term = mag === 1 ? s : `${mag}${s}`;
      if (!parts.length) {
        parts.push(c > 0 ? term : `-${term}`);
      } else {
        parts.push(`${sign}${term}`);
      }
    };

    add(this.b, "i");
    add(this.c, "j");
    add(this.d, "k");

    return parts.length ? parts.join("") : "0";
  }
}

File: code/swift/exercises.swift
import Foundation

public func firstThenLowerCase(of array: [String], satisfying p: (String) -> Bool) -> String? {
    array.first(where: p)?.lowercased()
}
public struct PhraseBuilder {
    private let words: [String]
    fileprivate init(_ words: [String]) { self.words = words }
    public func and(_ word: String) -> PhraseBuilder { PhraseBuilder(words + [word]) }
    public var phrase: String { words.joined(separator: " ") }
}
public func say() -> PhraseBuilder { PhraseBuilder([]) }
public func say(_ word: String) -> PhraseBuilder { PhraseBuilder([word]) }
public func meaningfulLineCount(_ filename: String) async -> Result<Int, Error> {
    do {
        let text = try String(contentsOfFile: filename, encoding: .utf8)
        let n = text.split(separator: "\n", omittingEmptySubsequences: false).filter { line in
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            return !trimmed.isEmpty && !trimmed.hasPrefix("#")
        }.count
        return .success(n)
    } catch {
        return .failure(error)
    }
}
public struct Quaternion: Equatable, CustomStringConvertible {
    public let a: Double, b: Double, c: Double, d: Double
    public static let ZERO = Quaternion(0,0,0,0)
    public static let I = Quaternion(0,1,0,0)
    public static let J = Quaternion(0,0,1,0)
    public static let K = Quaternion(0,0,0,1)
    public init(_ a: Double, _ b: Double, _ c: Double, _ d: Double) {
        self.a = a; self.b = b; self.c = c; self.d = d
    }
    public init(a: Double = 0, b: Double = 0, c: Double = 0, d: Double = 0) {
        self.init(a, b, c, d)
    }
    public var conjugate: Quaternion { Quaternion(a, -b, -c, -d) }
    public var coefficients: [Double] { [a, b, c, d] }
    public static func +(x: Quaternion, y: Quaternion) -> Quaternion {
        Quaternion(x.a + y.a, x.b + y.b, x.c + y.c, x.d + y.d)
    }
    public static func *(x: Quaternion, y: Quaternion) -> Quaternion {
        Quaternion(x.a*y.a - x.b*y.b - x.c*y.c - x.d*y.d,
                   x.a*y.b + x.b*y.a + x.c*y.d - x.d*y.c,
                   x.a*y.c - x.b*y.d + x.c*y.a + x.d*y.b,
                   x.a*y.d + x.b*y.c - x.c*y.b + x.d*y.a
        )  
    }
    public var description: String {
        if a == 0 && b == 0 && c == 0 && d == 0 { return "0" }
        var parts: [String] = []
        if a != 0 { parts.append("\(a)") }
        func term(_ coef: Double, _ sym: String) {
            if coef == 0 { return }
            let sign = coef < 0 ? "-" : (parts.isEmpty ? "" : "+")
            let mag = abs(coef)
            if mag == 1 { parts.append("\(sign)\(sym)") }
            else { parts.append("\(sign)\(mag)\(sym)") }
        }
        term(b, "i"); term(c, "j"); term(d, "k")
        return parts.joined()
    }
}
public indirect enum BinarySearchTree {
    case empty
    case node(String, BinarySearchTree, BinarySearchTree)
    public static var Empty: BinarySearchTree { .empty }
    public func insert(_ value: String) -> BinarySearchTree {
        switch self {
            case .empty: return .node(value, .empty, .empty)
            case let .node(v, l, r):
                if value < v { return .node(v, l.insert(value), r) }
                if value > v { return .node(v, l, r.insert(value)) }
                return self
        }
    }
    public func contains(_ value: String) -> Bool {
        switch self {
            case .empty: return false
            case let .node(v, l, r):
                if value == v { return true }
                return value < v ? l.contains(value) : r.contains(value)
        }
    }
    public var size: Int {
        switch self {
            case .empty: return 0
            case let .node(_, l, r): return 1 + l.size + r.size
        }
    }
}
extension BinarySearchTree: CustomStringConvertible {
    public var description: String {
        switch self {
        case .empty:
            return "()"
        case let .node(v, l, r):
            var s = "("
            if case .empty = l {} else { s += l.description }
            s += v
            if case .empty = r {} else { s += r.description }
            s += ")"
            return s
        }
    }
}

